### ISSUE TITLE
Wrong return type `iterable` of `UrlProviderInterface::generate`

### ISSUE THREAD

**Issue by mbenoukaiss** [stefandoorn/sitemap-plugin#258]
`UrlProviderInterface::generate` is typed as returning an iterable which I believe is wrong, an array is iterable, but a Generator is also an iterable. However the value returned by generate is sent to `array_merge` which only accepts arrays.

Seeing the return type of `UrlProviderInterface::generate` made me believe I could use yield and return a generator in my implementation but it is not the case and will give the error `array_merge(): Argument #1 must be of type array, Generator given`.

I see three possible solutions : 
- Change `UrlProviderInterface::generate` signature to return array
- Call `iterator_to_array` in `SitemapBuilder::generate` if the value is not an array
- Use `AppendIterator` instead of merging arrays to make use of the iterators returned, which if used properly could lead to less memory usage since we don't have to load the whole URLs array in memory (however I don't know about CPU performance). But is it worth for a command ? I don't know much about the codebase but it could make it more complex for negligible gains.

Whatever your decision is I can gladly make a PR

**Comment by mbenoukaiss** [comment]
Yep, I will try to look into it tomorrow!

**Comment by mbenoukaiss** [comment]
Yep, I will try to look into it tomorrow, just tell me which branch to start from!

**Comment by stefandoorn** [comment]
Nice!

Branch: https://github.com/stefandoorn/sitemap-plugin/tree/versions/2.4

**Comment by stefandoorn** [comment]
Thanks!

https://github.com/stefandoorn/sitemap-plugin/releases/tag/v2.4.3
